addons/purchase/models/purchase.py

```diff
@@ -200,7 +200,7 @@ def onchange_partner_id(self):
             self.payment_term_id = False
             self.currency_id = self.env.company.currency_id.id
         else:
-            self.fiscal_position_id = self.env['account.fiscal.position'].with_context(company_id=self.company_id.id).get_fiscal_position(self.partner_id.id)
+            self.fiscal_position_id = self.env['account.fiscal.position'].with_company(self.company_id).get_fiscal_position(self.partner_id.id)
             self.payment_term_id = self.partner_id.property_supplier_payment_term_id.id
```

Comment:
TODO access property_supplier_payment_term with correct company, same for property_purchase_currency on next line